lib/methods/bulk-action.ts

```diff
@@ -0,0 +1,121 @@
+import delay from 'delay'
+import {
+  BulkAction,
+  BulkActionPayload,
+  BulkActionProps,
+  BulkActionStatus,
+} from '../entities/bulk-action'
+import { PlainClientAPI } from '../plain/common-types'
+
+const DEFAULT_MAX_RETRIES = 30
+const DEFAULT_INITIAL_DELAY_MS = 1000
+const DEFAULT_RETRY_INTERVAL_MS = 2000
+
+export class BulkActionProcessingError extends Error {
+  public bulkAction?: BulkAction | BulkActionProps
+
+  constructor(message: string, bulkAction?: BulkAction | BulkActionProps) {
+    super(message)
+    this.bulkAction = bulkAction
+    this.name = this.constructor.name
+  }
+}
+
+export class BulkActionFailedError extends BulkActionProcessingError {}
+
+export type BulkActionProcessingOptions = {
+  /** The amount of times to retry.
+   * @default 30
+   * */
+  retryCount?: number
+  /** The interval between retries, in milliseconds (ms).
+   * @default 2000 (2s)
+   * */
+  retryIntervalMs?: number
+  /**
+   * Initial delay in milliseconds when performing the first check.
+   * This is used to prevent short running bulkActions of waiting too long for a result.
+   * @default 1000 (1s)
+   * */
+  initialDelayMs?: number
+  /**
+   * Throws an error if the BulkAction does not complete with a successful (succeeded) status.
+   * @default false
+   */
+  throwOnFailedExecution?: boolean
```

Comment:
Are we sure we want that to default to false?

Reply:
It can be the opposite, might be actually better ^.